// @alien_org/contract/dist/index.mjs
//#region src/launch-params.ts
/**
* Supported platforms for miniapps.
*/
const PLATFORMS = ["ios", "android"];

//#endregion
//#region src/methods/versions/releases.ts
const releases = {
	"0.0.9": ["app:ready"],
	"0.0.14": ["miniapp:close.ack", "host.back.button:toggle"],
	"0.1.1": [
		"payment:request",
		"clipboard:write",
		"clipboard:read"
	],
	"0.1.3": ["link:open"]
};

//#endregion
//#region src/methods/versions/get-release-version.ts
function getReleaseVersion(method, payload) {
	return Object.keys(releases).find((version) => {
		const releaseItems = releases[version];
		if (!releaseItems) return false;
		return releaseItems.some((item) => {
			if (payload) return typeof item === "object" && item.method === method && item.param === payload;
			return item === method;
		});
	}) || null;
}

//#endregion
//#region src/methods/versions/index.ts
/**
* Compare two semver versions.
* @returns negative if a < b, 0 if a === b, positive if a > b
*/
function compareVersions(a, b) {
	const [aMajor, aMinor, aPatch] = a.split(".").map(Number);
	const [bMajor, bMinor, bPatch] = b.split(".").map(Number);
	if (aMajor !== bMajor) return (aMajor ?? 0) - (bMajor ?? 0);
	if (aMinor !== bMinor) return (aMinor ?? 0) - (bMinor ?? 0);
	return (aPatch ?? 0) - (bPatch ?? 0);
}
/**
* Check if a method is supported in a given version.
*
* Uses the minimum version that introduced the method and returns true if
* the given version is >= that minimum (semver comparison). This supports
* versions not explicitly listed in releases (e.g. 0.1.4 when method was
* added in 0.1.1).
*
* @param method - The method name to check.
* @param version - The contract version (must be a valid version string, not undefined).
* @returns `true` if the method is supported in the given version, `false` otherwise.
*
* @remarks
* This function only accepts valid version strings. Version existence checks should be
* handled at a higher level before calling this function.
*/
function isMethodSupported(method, version) {
	const minVersion = getMethodMinVersion(method);
	if (!minVersion) return false;
	return compareVersions(version, minVersion) >= 0;
}
/**
* Get the minimum version that supports a method.
* Returns undefined if method not found in any version.
*/
function getMethodMinVersion(method) {
	const sorted = Object.keys(releases).sort(compareVersions);
	for (const version of sorted) {
		const methods = releases[version];
		if (!methods) continue;
		if (methods.some((m) => typeof m === "string" ? m === method : m.method === method)) return version;
	}
}

//#endregion
export { PLATFORMS, getMethodMinVersion, getReleaseVersion, isMethodSupported, releases };
